src/algebra/module_regular.lean

```diff
@@ -0,0 +1,168 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.group
+import algebra.group_power.basic
+import algebra.iterate_hom
+import algebra.module.basic
+
+/-!
+# Action of regular elements on a module
+
+We introduce `M`-regular elements, in the context of an `R`-module `M`.
+
+By definition, an `M`-regular element of a commutative ring acting on an `R`-module `M` is an
+element `a ∈ R` such that the function `m ↦ a • m` is injective.
+-/
+
+--should it be in the module namespace?
+
+namespace module
+
+variables {R : Type*} {a b : R} (M : Type*)
```

Comment:
I think you should make this a statement over different types, and then the `is_scalar_tower` below makes more sense
```suggestion
variables {R S : Type*} {ar br : R} {as bs : S} (M : Type*)
```

Reply:
Eric, I am looking at this.  Most proofs work in the more general context, except when I want to multiply elements in the two possible orders.  This can actually make sense, even the the intermediate ring `S`, but it requires being careful with the bracketing and I am not convinced that it can be done at the level of multiplying by elements.

I am thinking of dropping `S` in those cases and leaving it with `mul_action.is_scalar_left`, once I get it working!

Reply:
That sounds fine, and it's much easier to discuss in difficult generalizations once the simple ones are done!

Reply:
I still think that there is some weirdness between `mul_action` and `is_scalar_tower`.  I am having a hard time finding the right combination of hypotheses that will allow me to prove the results with units.

Reply:
I am going to bed now.  I pushed a version for which the linter on my computer seemed happy.  I had lots of duplicate assumptions, so this might explain the quirks in the proofs.  I am not sure that this will work, but at least I removed a lot unneeded assumptions!